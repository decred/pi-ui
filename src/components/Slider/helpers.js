export const DIMENSIONS_MAP = {
  x: "width",
  y: "height"
};

export const POSITIONS_MAP = {
  x: "left",
  y: "top"
};

export const POSITIONS_MAP_CAPITALIZED = {
  x: "Left",
  y: "Top"
};

export function getClientPosition(e) {
  return e.touches?.length
    ? {
        x: e.touches[0].clientX,
        y: e.touches[0].clientY
      }
    : {
        x: e.clientX,
        y: e.clientY
      };
}

export function addEventListeners(handleDrag, handleDragEnd) {
  document.addEventListener("mousemove", handleDrag);
  document.addEventListener("touchmove", handleDrag, { passive: false });

  document.addEventListener("mouseup", handleDragEnd);
  document.addEventListener("touchend", handleDragEnd);
  document.addEventListener("touchcancel", handleDragEnd);
}
