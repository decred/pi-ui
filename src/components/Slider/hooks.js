import { useCallback, useEffect, useMemo, useRef } from "react";
import {
  getClientPosition,
  addEventListeners,
  DIMENSIONS_MAP,
  POSITIONS_MAP,
  POSITIONS_MAP_CAPITALIZED
} from "./helpers";

function useSliderHandle(
  container,
  double,
  disabled,
  axis,
  min,
  max,
  step,
  value,
  onChange,
  onDragStart,
  onDragEnd,
  barrier
) {
  const handle = useRef(null);
  const start = useRef();
  const offset = useRef();

  useEffect(() => {
    if (value > max) onChange(max);
    if (value < min) onChange(min);
  }, [value, min, max, onChange]);

  if (double && !barrier(value)) {
    onChange(value - step);
  }

  const position = useMemo(() => {
    let newValue = ((value - min) / (max - min)) * 100;

    if (newValue > 100) newValue = 100;
    if (newValue < 0) newValue = 0;

    return axis === "x"
      ? { top: 0, left: newValue }
      : { top: newValue, left: 0 };
  }, [value, min, max, axis]);

  const change = (position) => {
    if (!onChange) return;

    const dimension = container.current.getBoundingClientRect()[
      DIMENSIONS_MAP[axis]
    ];
    let ds = 0;

    if (position < 0) position = 0;
    if (position > dimension) position = dimension;

    ds = (position / dimension) * (max - min);

    const newPosition = (ds !== 0 ? parseInt(ds / step, 10) * step : 0) + min;

    if (!double || barrier(newPosition)) {
      onChange(newPosition);
    }
  };

  const getPos = (e) => {
    const clientPos = getClientPosition(e)[axis];
    const position = clientPos + start.current - offset.current;
    return position;
  };

  const handleDrag = (e) => {
    if (disabled) return;

    e.preventDefault();
    change(getPos(e));
  };

  const handleDragEnd = (e) => {
    if (disabled) return;

    e.preventDefault();
    document.removeEventListener("mousemove", handleDrag);
    document.removeEventListener("touchmove", handleDrag, {
      passive: false
    });

    document.removeEventListener("mouseup", handleDragEnd);
    document.removeEventListener("touchend", handleDragEnd);
    document.removeEventListener("touchcancel", handleDragEnd);

    if (onDragEnd) {
      onDragEnd(e);
    }
  };

  const handleMouseDown = (e) => {
    if (disabled) return;

    e.preventDefault();
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();

    const dom = handle.current;
    const clientPos = getClientPosition(e)[axis];

    start.current = dom[`offset${POSITIONS_MAP_CAPITALIZED[axis]}`];
    offset.current = clientPos;

    addEventListeners(handleDrag, handleDragEnd);

    if (onDragStart) {
      onDragStart(e);
    }
  };

  return {
    position,
    handleMouseDown,
    handle,
    handleDrag,
    handleDragEnd,
    change
  };
}

function useSlider(double, disabled, axis, min, max, step, handles) {
  const container = useRef(null);

  if (double) {
    const [leftHandle, rightHandle] = handles;
    leftHandle.barrier = useCallback((value) => value <= rightHandle.value, [
      rightHandle.value
    ]);
    rightHandle.barrier = useCallback((value) => value >= leftHandle.value, [
      leftHandle.value
    ]);
  }

  const handleHooks = handles.map((handle) =>
    useSliderHandle(
      container,
      double,
      disabled,
      axis,
      min,
      max,
      step,
      handle.value,
      handle.onChange,
      handle.onDragStart,
      handle.onDragEnd,
      handle.barrier
    )
  );

  const handleTrackMouseDown = (e) => {
    if (disabled) return;

    e.preventDefault();
    const clientPos = getClientPosition(e)[axis];
    const rect = container.current.getBoundingClientRect();

    let handleIndex = 0;

    if (double) {
      const [leftHandlePosition, rightHandlePosition] = handleHooks.map(
        (handleHook) => {
          const rect = handleHook.handle.current.getBoundingClientRect();
          return rect[axis] + rect[DIMENSIONS_MAP[axis]] / 2;
        }
      );

      if (
        (clientPos > leftHandlePosition &&
          clientPos < rightHandlePosition &&
          clientPos - leftHandlePosition >= rightHandlePosition - clientPos) ||
        clientPos >= rightHandlePosition
      )
        handleIndex = 1;
    }

    addEventListeners(
      handleHooks[handleIndex].handleDrag,
      handleHooks[handleIndex].handleDragEnd
    );

    handleHooks[handleIndex].change(clientPos - rect[POSITIONS_MAP[axis]]);

    if (handleHooks[handleIndex].onDragStart) {
      handleHooks[handleIndex].onDragStart(e);
    }
  };

  const valueStyle = {};

  const position = POSITIONS_MAP[axis];

  const leftHandlePositionAtAxis = handleHooks[0].position[position];

  if (double) {
    const rightHandlePositionAtAxis = handleHooks[1].position[position];

    valueStyle[position] = leftHandlePositionAtAxis + "%";

    valueStyle[DIMENSIONS_MAP[axis]] =
      rightHandlePositionAtAxis - leftHandlePositionAtAxis + "%";
  } else {
    valueStyle[DIMENSIONS_MAP[axis]] = leftHandlePositionAtAxis + "%";
  }

  return {
    container,
    handleTrackMouseDown,
    valueStyle,
    handleHooks
  };
}

export default useSlider;
