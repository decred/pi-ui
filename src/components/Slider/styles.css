.thumb {
  position: relative;
  display: block;
  content: "";
  width: 1.8rem;
  height: 1.8rem;
  background-color: var(--main-color);
  border-radius: 50%;
  box-shadow: 0 0.1rem 0.1rem rgba(0, 0, 0, 0.5);
  user-select: none;
  cursor: pointer;
  box-sizing: border-box;
}

.track {
  position: relative;
  display: inline-block;
  background-color: var(--btn-disabled-background-color);
  border-radius: 0.5rem;
  user-select: none;
  box-sizing: border-box;
}

.x.track {
  width: 20rem;
  height: 1rem;
}

.y.track {
  width: 1rem;
  height: 20rem;
}

.active {
  position: absolute;
  background-color: var(--btn-background-color);
  border-radius: 0.5rem;
  user-select: none;
  box-sizing: border-box;
}

.x.active {
  top: 0;
  height: 100%;
}

.y.active {
  left: 0;
  width: 100%;
}

.disabled {
  opacity: 0.5;
}

.handle {
  z-index: 1;
  position: absolute;
  transform: translate(-50%, -50%);
}
