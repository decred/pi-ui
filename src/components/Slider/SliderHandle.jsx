import React from "react";
import PropTypes from "prop-types";
import { POSITIONS_MAP } from "./helpers";
import { classNames } from "../../utils";
import styles from "./styles.css";

const SliderHandle = React.forwardRef(
  ({ axis, position, onTouchStart, onMouseDown, className }, ref) => {
    const style = { top: "50%", left: "50%" };
    style[POSITIONS_MAP[axis]] = position[POSITIONS_MAP[axis]] + "%";

    return (
      <div
        {...{ ref, style, onTouchStart, onMouseDown }}
        className={styles.handle}
        onClick={(e) => {
          e.stopPropagation();
          e.nativeEvent.stopImmediatePropagation();
        }}>
        <div className={classNames(className, styles.thumb)} />
      </div>
    );
  }
);

SliderHandle.propTypes = {
  axis: PropTypes.string.isRequired,
  position: PropTypes.object.isRequired,
  style: PropTypes.object,
  onTouchStart: PropTypes.func.isRequired,
  onMouseDown: PropTypes.func.isRequired,
  className: PropTypes.string
};

export default SliderHandle;
