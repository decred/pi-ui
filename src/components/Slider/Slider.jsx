import React from "react";
import PropTypes from "prop-types";
import { classNames } from "../../utils";
import SliderHandle from "./SliderHandle.jsx";
import useSlider from "./hooks";
import styles from "./styles.module.css";

const Slider = ({
  double,
  disabled,
  axis,
  min,
  max,
  step,
  handles,
  className,
  rangeClassName,
  ...props
}) => {
  const { container, handleTrackMouseDown, valueStyle, handleHooks } =
    useSlider(double, disabled, axis, min, max, step, handles);

  const [leftHandleHook, rightHandleHook] = handleHooks;

  return (
    <div
      ref={container}
      className={classNames(
        styles[axis],
        styles.track,
        disabled && styles.disabled,
        className
      )}
      onTouchStart={handleTrackMouseDown}
      onMouseDown={handleTrackMouseDown}
      data-testid="slider-body"
      {...props}>
      {double && (
        <SliderHandle
          ref={rightHandleHook.handle}
          axis={axis}
          position={rightHandleHook.position}
          onTouchStart={rightHandleHook.handleMouseDown}
          onMouseDown={rightHandleHook.handleMouseDown}
          className={handles[1].className}
          dataTestId="slider-handle-1"
        />
      )}
      <div
        className={classNames(styles[axis], styles.active, rangeClassName)}
        style={valueStyle}
        data-testid="slider-bar"
      />
      <SliderHandle
        ref={leftHandleHook.handle}
        axis={axis}
        position={leftHandleHook.position}
        onTouchStart={leftHandleHook.handleMouseDown}
        onMouseDown={leftHandleHook.handleMouseDown}
        className={handles[0].className}
        dataTestId="slider-handle-0"
      />
    </div>
  );
};

Slider.propTypes = {
  double: PropTypes.bool,
  disabled: PropTypes.bool,
  axis: PropTypes.string,
  max: PropTypes.number,
  min: PropTypes.number,
  step: PropTypes.number,
  handles: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.number,
      onChange: PropTypes.func,
      onDragStart: PropTypes.func,
      onDragEnd: PropTypes.func,
      className: PropTypes.string,
    })
  ),
  className: PropTypes.string,
  rangeClassName: PropTypes.string,
};

Slider.defaultProps = {
  double: false,
  disabled: false,
  axis: "x",
  min: 0,
  max: 100,
  step: 1,
  handles: [{ value: 0 }, { value: 40 }],
  className: "",
};

export default Slider;
