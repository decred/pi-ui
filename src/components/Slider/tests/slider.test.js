import React from "react";
import { create } from "react-test-renderer";
import Slider from "../Slider";

describe("Slider component", () => {
  test("Matches snapshot", () => {
    let value0 = 0;
    let value1 = 40;
    const mockHandleChange0 = jest.fn((newValue) => {
      value0 = newValue;
    });
    const mockHandleChange1 = jest.fn((newValue) => {
      value1 = newValue;
    });
    const slider = create(
      <Slider
        double={true}
        axis="x"
        max={100}
        min={0}
        handles={[
          {
            value: value0,
            onChange: mockHandleChange0
          },
          {
            value: value1,
            onChange: mockHandleChange1
          }
        ]}
      />
    );
    expect(slider.toJSON()).toMatchSnapshot();
  });
});
