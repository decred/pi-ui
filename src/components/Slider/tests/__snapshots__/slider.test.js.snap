// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component Matches snapshot 1`] = `
<div
  className="x track"
  data-testid="slider-body"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="handle"
    data-testid="slider-handle-1"
    onClick={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "40%",
        "top": "50%",
      }
    }
  >
    <div
      className="thumb"
    />
  </div>
  <div
    className="x active"
    data-testid="slider-bar"
    style={
      Object {
        "left": "0%",
        "width": "40%",
      }
    }
  />
  <div
    className="handle"
    data-testid="slider-handle-0"
    onClick={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "0%",
        "top": "50%",
      }
    }
  >
    <div
      className="thumb"
    />
  </div>
</div>
`;
